content/blog/2015-05-24-custom-sd.md

```diff
@@ -0,0 +1,311 @@
+
+1. Extract your scrape targets from your service discovery
+2. Transform them into groups of targets Prometheus can read
+3. Write them to disk for Prometheus to read
+
+We actually skipped step 0 — which is to tell Prometheus where those files will be
+located.
+
+A popular way to do service discovery are distributed key-value stores. As they put no
+constraints on how to do service discovery, a custom bridge is an efficient approach to
+overcome Prometheus's lack of knowledge about your setup.
+
+
+## SD bridge by example
+
+Let's get started with a fictional setup of an etcd cluster in which our services and
+their instances are stored in the following format:
+
+```
+/services
+    /service-x
+        /info         = { "owner": "prodeng", "monitored": true }
+        /instances
+            /inst-1   = { "host": "10.33.1.230", "port": 5441, "group": "production" }
+            ...
+            /inst-n   = { "host": "10.33.1.244", "port": 8223, "group": "canary" }
+
+    /service-y
+        ...
+```
+
+Our keys contain JSON-serialized data. For the `*/info` keys this a simple map of labels
+that provide metadata about the service.
+Each instance is a JSON object containing the host and port of the instance, as well as
+the a group it belongs to within the service.
+
+Our goal is to make Prometheus scrape all services, that are flagged for monitoring and
+assign the correct group and owner labels to all targets.
+
+
+### Step 1 — Extract
+
+First, we want to write a simple Go program that continiously watches etcd for changes in our 
+`/services` directory.
+
+We define data structures that map to our stored instances and service information:
+
+```
+// Instance is the instance object stored in etcd.
+type Instance struct {
+	Host  string `json:"host"`
+	Port  int    `json:"port"`
+	Group string `json:"group"`
+}
+
+// ServiceInfo is the information object stored in etcd.
+type ServiceInfo struct {
+	Owner     string `json:"owner"`
+	Monitored bool   `json:"monitored"`
+}
+```
+
+The current state of our service discovery is stored inside our program in the following structs.
+
+```
+// service is a service stored in etcd.
+type service struct {
+	info      *ServiceInfo
+	instances map[string]*Instance // Instances by their path.
+}
+
+// services are the services stored in etcd.
+type services struct {
+	m   map[string]*service // The current services by name.
+	del []string            // Services deleted in the last update.
+}
+```
+
+Using the etcd client library we now perform one initial scan of the `/services` directory 
+and populate our just defined structs. We then proceed by watching the whole directory
+recursively.
+Whenever we receive a setting or deleting operation on one of the nodes, we update our state.
+Each time our state changes we persist it to disk in a Prometheus-readbable format.
+
+This procedure results in the following `main()` function for our SD bridge:
+
+```
+func main() {
```

Comment:
@matthiasr suggested that we can make things simple by doing the example with `confd`. That's an option I'll look into later. It could likely be a lot shorter without being unrealistic.

Hope we'll find a middle ground that way.